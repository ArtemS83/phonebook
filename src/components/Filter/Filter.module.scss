.title {
  margin-top: 10px;
  font-size: 18px;
  font-weight: 500;
  color: #494141;
}
.input {
  margin-top: 5px;
  padding: 20px;
  width: 100%;
  height: 30px;
  font-size: 18px;
  background-color: rgba(239, 235, 241, 0.883);
  border: 1px solid rgb(125, 19, 196);
  border-radius: 4px;
  outline: none;
  box-shadow: 2px 6px 8px -2px #a09c9c;
  cursor: pointer;
  transition: border 300ms ease-in-out;
  &:focus,
  &:hover {
    border: 2px solid rgb(154, 13, 161);
  }
}
