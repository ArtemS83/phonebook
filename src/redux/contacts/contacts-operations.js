import axios from 'axios';
import {
  addContactRequest,
  addContactSuccess,
  addContactError,
  fetchContactsRequest,
  fetchContactsSuccess,
  fetchContactsError,
  deleteContactRequest,
  deleteContactSuccess,
  deleteContactError,
  updateContactRequest,
  updateContactSuccess,
  updateContactError,
} from '../../redux/contacts/contacts-actions';

export const fetchContacts = () => dispatch => {
  dispatch(fetchContactsRequest());
  axios
    .get('/contacts')
    .then(({ data }) => dispatch(fetchContactsSuccess(data)))
    .catch(error => dispatch(fetchContactsError(error.message)));
};

export const addContact = (name, number) => dispatch => {
  const contact = {
    name,
    number,
  };
  dispatch(addContactRequest());
  axios
    .post('/contacts', contact)
    .then(({ data }) => dispatch(addContactSuccess(data)))
    .catch(error => dispatch(addContactError(error.message)));
};

export const deleteContact = id => dispatch => {
  dispatch(deleteContactRequest());
  axios
    .delete(`/contacts/${id}`)
    .then(() => dispatch(deleteContactSuccess(id)))
    .catch(error => dispatch(deleteContactError(error.message)));
};

export const updateContact = (id, name, number) => dispatch => {
  const contact = {
    name,
    number,
  };
  dispatch(updateContactRequest());
  axios
    .patch(`/contacts/${id}`, contact)
    .then(({ data }) => {
      dispatch(updateContactSuccess(data));
    })
    .catch(error => dispatch(updateContactError(error.message)));
};
