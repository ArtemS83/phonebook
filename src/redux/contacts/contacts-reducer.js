import { createReducer, combineReducers } from '@reduxjs/toolkit';
import {
  addContactRequest,
  addContactSuccess,
  addContactError,
  fetchContactsRequest,
  fetchContactsSuccess,
  fetchContactsError,
  filterContacts,
  deleteContactRequest,
  deleteContactSuccess,
  deleteContactError,
  updateContactRequest,
  updateContactSuccess,
  updateContactError,
} from '../../redux/contacts/contacts-actions';
import authActions from '../../redux/auth/auth-actions';

const initialState = [];
const items = createReducer(initialState, {
  [fetchContactsSuccess]: (_, action) => action.payload,
  [addContactSuccess]: (state, action) => [...state, action.payload],
  [deleteContactSuccess]: (state, action) =>
    state.filter(({ id }) => id !== action.payload),
  [updateContactSuccess]: (state, action) =>
    state.map(contact =>
      contact.id === action.payload.id ? action.payload : contact,
    ),
  [authActions.logoutSuccess]: () => [],
});

const filter = createReducer('', {
  [filterContacts]: (_, action) => action.payload.trim(),
});

const loading = createReducer(false, {
  [fetchContactsRequest]: () => true,
  [fetchContactsSuccess]: () => false,
  [fetchContactsError]: () => false,

  [addContactRequest]: () => true,
  [addContactSuccess]: () => false,
  [addContactError]: () => false,

  [deleteContactRequest]: () => true,
  [deleteContactSuccess]: () => false,
  [deleteContactError]: () => false,

  [updateContactRequest]: () => true,
  [updateContactSuccess]: () => false,
  [updateContactError]: () => false,
});

const setError = (_, { payload }) => payload;

const error = createReducer(null, {
  [fetchContactsError]: setError,
  [addContactError]: setError,
  [deleteContactError]: setError,
  [updateContactError]: setError,
});

export default combineReducers({
  items,
  filter,
  loading,
  error,
});
