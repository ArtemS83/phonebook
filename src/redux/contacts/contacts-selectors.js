import { createSelector } from '@reduxjs/toolkit';

export const getContacts = state => state.contacts.items;

export const getFilter = state => state.contacts.filter;
// проверяет изменился ли state(items и filter), чтобы без надобности не рендерить
export const getVisibleContacts = createSelector(
  [getContacts, getFilter],
  (contacts, filter) => {
    const normalizedFilter = filter.toLowerCase();
    return contacts.filter(contact =>
      contact.name.toLowerCase().includes(normalizedFilter),
    );
  },
);

export const getIsLoadingSelector = state => state.contacts.loading;

// export const getVisibleContacts = state => {
//   const contacts = getContacts(state);
//   const filter = getFilter(state);
//   const normalizedFilter = filter.toLowerCase();

//   return contacts.filter(contact =>
//     contact.name.toLowerCase().includes(normalizedFilter),
//   );
// };
