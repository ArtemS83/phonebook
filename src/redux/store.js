import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import contactsReducer from './contacts/contacts-reducer';
import authReducer from './auth/auth-reducer';

const authPersistConfig = {
  key: 'token',
  version: 1,
  storage,
  whitelist: ['token'],
};

const store = configureStore({
  reducer: {
    contacts: contactsReducer,
    auth: persistReducer(authPersistConfig, authReducer),
  },
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  // devTools: process.env.NODE_ENV === 'development',
});

const persistor = persistStore(store);

export default { store, persistor }; // eslint-disable-line
