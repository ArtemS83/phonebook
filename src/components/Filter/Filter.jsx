import PropTypes from 'prop-types';
import { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { filterContacts } from '../../redux/contacts/contacts-actions';
import { getFilter } from '../../redux/contacts/contacts-selectors';
import style from './Filter.module.scss';

const Filter = ({ text }) => {
  const filter = useSelector(getFilter);
  const dispatch = useDispatch();

  const handleInputFindChange = useCallback(
    ({ target }) => {
      const { value } = target;
      dispatch(filterContacts(value));
    },
    [dispatch],
  );
  // const handleInputFindChange = ({ target }) => {
  //   const { value } = target;
  //   dispatch(filterContacts(value));
  // };

  return (
    <>
      <p data-testid="filtertextTest" className={style.title}>
        {text}
      </p>
      <input
        data-testid="filterInputTest"
        className={style.input}
        type="text"
        name="filter"
        value={filter}
        placeholder="Enter name contact"
        onChange={handleInputFindChange}
      />
    </>
  );
};

Filter.propTypes = {
  text: PropTypes.string.isRequired,
};

export default Filter;
