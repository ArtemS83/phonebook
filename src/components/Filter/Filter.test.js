import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../redux/store';
import { filterContacts } from '../../redux/contacts/contacts-actions';
import reducer from '../../redux/contacts/contacts-reducer';
import Filter from './Filter';

describe('Filter', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={store.store}>
        <Filter text="TEXT render" />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('handleInputFindChange should return a new object', () => {
    expect(filterContacts('Test change text')).toMatchObject({
      type: 'contacts/filter',
      payload: 'Test change text',
    });
  });

  it('reducer-loading', () => {
    const initialState = {
      items: [],
      filter: 'AAA',
      loading: false,
      error: null,
    };

    const action = { type: 'contacts/addContactRequest' };
    const newState = reducer(initialState, action);

    expect(newState).toMatchObject({
      items: [],
      filter: 'AAA',
      loading: true,
      error: null,
    });
  });

  it('reducer-filter', () => {
    const initialState = {
      items: [],
      filter: 'BBB',
      loading: false,
      error: null,
    };

    const action = { type: 'contacts/filter', payload: 'AAA' };
    const newState = reducer(initialState, action);

    expect(newState).toMatchObject({
      items: [],
      filter: 'AAA',
      loading: false,
      error: null,
    });
  });
});
