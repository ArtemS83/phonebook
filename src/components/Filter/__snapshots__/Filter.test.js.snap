// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should render 1`] = `
<div>
  <p
    class="title"
    data-testid="filtertextTest"
  >
    TEXT render
  </p>
  <input
    class="input"
    data-testid="filterInputTest"
    name="filter"
    placeholder="Enter name contact"
    type="text"
    value=""
  />
</div>
`;
